### DISCUSSION TITLE
Modifying serialization of fields based on other field (in v2)

### DISCUSSION THREAD

**Discussion by Lithimlin** [pydantic/pydantic#6296] [Question] [answered]
I'm currently trying to serialize some metrics I get via a json. That data is nested fairly deeply and on the deepest levels, there are some relative timestamps represented by `timedelta`s. However, the base timestamp is located on the first level of the model. How would I go about applying this base time to all (relevant) time deltas? If possible I don't want to pass the timestamp to every other model on the way to the ones that actually need it. Below is an example structure:

```py
from datetime import datetime, timedelta
from typing import Sequence
from pydantic import BaseModel

class Interval(BaseModel):
    start: timedelta # combine with timestamp
    end: timedelta # combine with timestamp
    duration: timedelta  # do not overwrite this field

class Foo(BaseModel):
    x: int
    intervals: Sequence[Interval]
    # I don't want to have to pass the timestamp here

class MetricsModel(BaseModel):
    timestamp: datetime
    foo: Foo
```

Note: I am using the new version 2 of pydantic

**Comment by samuelcolvin** [comment]
I'm a bit confused about the input data and the expected calculations, so hard to answer.

You could use either:
* Model validators - https://docs.pydantic.dev/dev-v2/usage/validators/#model-validators
* or AliasPath - https://docs.pydantic.dev/dev-v2/usage/fields/#aliaspath-and-aliaschoices

Or a combination of both.

**Reply by Lithimlin** [reply]
Some exemplary input data for the MWE above may look like this:
```json
{
  "timestamp": 1688008832,
  "foo": {
    "x": 42,
    "intervals": [
      {
        "start": 0.00,
        "end": 3.5694,
        "duration": 3.5694,
      },
      {
        "start": 3.602,
        "end": 3.807
        "duration": 0.205,
      },
    ],
  },
}
```
In reality, I'm working with a JSON of the output from iperf3.
So far I haven't stumbled across AliasPaths yet so I'll check that out.
I don't think I want to use a validator here as I don't actually want to store the converted timestamps. I only want to convert them when I dump the model somewhere. I already got the part of getting the data into the model, now I want to get it out again when I save it somewhere and in that case, I want to convert the `timedelta` to the appropriate `datetime`.

**Reply by Lithimlin** [reply]
I've taken a look at AliasPaths and I don't see how they would be helpful here

**Reply by samuelcolvin** [reply]
That data looks like it matches the model, I don't really see what's wrong?

**Reply by Lithimlin** [reply]
There's nothing wrong with the parsing.
I just want to automatically convert the `start` and `end` deltas to timestamps when dumping the model.

**Reply by samuelcolvin** [reply]
Oh sorry, I misread the issue. You probably want to user customer serializers - https://docs.pydantic.dev/dev-v2/usage/exporting_models/#custom-serializers

**Reply by Lithimlin** [reply]
Yes, that is in fact what I want, but how do I go about using the `timestamp` field from the `MetricsModel` in the serializer of my `Interval` model?

**Reply by samuelcolvin** [reply]
you would need to add a `model_serializer` for `MetricsModel` and thereby customuse how `foo.interals` is serialized.

In future, we might add `context` to serialization like we have on validation, which would provide another way pass the information to `Interval`.

**Reply by Lithimlin** [reply]
Sorry, I'm finally able to get back to this...
> you would need to add a `model_serializer` for `MetricsModel` and thereby customuse how `foo.interals` is serialized.

How would I go about doing this in the example above? How do I manipulate the serialization of `foo` from inside the `MetricsModel`'s serializer?

**Reply by Lithimlin** [reply, marked as answer]
For now I solved it by defining extra functions for `Foo` and `Interval` that do take a `datetime` as a base time, cascading it throughout the classes during serialization. However, this is really tedious when the structures become deeper, so an extra `context` field would be appreciated.

**Comment by Lithimlin** [comment]
As said in https://github.com/pydantic/pydantic/discussions/6296#discussioncomment-6315162, I've created a chain of serializers all the way down to the `start` and `end` fields of my `Interval`:

```py
from datetime import datetime, timedelta
from typing import Any, Sequence

from pydantic import BaseModel, model_serializer


class Interval(BaseModel):
    start: timedelta
    end: timedelta
    duration: timedelta  # do not overwrite this field

    def serialize_with_timestamp(self, timestamp: datetime) -> dict[str, Any]:
        return {
            "start": timestamp + self.start,
            "end": timestamp + self.end,
            "duration": self.duration,
        }


class Foo(BaseModel):
    x: int
    intervals: Sequence[Interval]

    def serialize_with_timestamp(self, timestamp: datetime) -> dict[str, Any]:
        return {
            "x": self.x,
            "intervals": [interval.serialize_with_timestamp(timestamp) for interval in self.intervals],
        }


class MetricsModel(BaseModel):
    timestamp: datetime
    foo: Foo

    @model_serializer
    def _serialize(self) -> dict[str, Any]:
        return {
            "timestamp": self.timestamp,
            "foo": self.foo,
        }
```

Hopefully in the future, there will be a more elegant solution to this.
